### DISCUSSION TITLE
Caching Selector Results

### DISCUSSION THREAD

**Discussion by david-trumid** [facebookexperimental/Recoil#992] [Q&A] [answered]
I am new to Recoil and from what I can tell the selector results may not be deep equal checked. Is this true, or is there an option for this? for now what I am doing is making a wrapper like this, but not sure if there is a concern with this (disregard the not handling of Promises, i just want to make sure im not missing something bigger that would cause issues)

```
export function selectorDeep<T>({ get, key }: ReadOnlySelectorOptions<T>): RecoilValueReadOnly<T> {
  let cache: T | Promise<T> | RecoilValue<T> | undefined;

  return selector({
    get: (getOptions) => {
      const newResult = get(getOptions);

      if (cache !== undefined && isEqual(cache, newResult)) {
        console.log(key, ' is equal ', cache, newResult);

        return cache;
      }

      console.log(key, ' is NOT equal ', cache, newResult);

      cache = newResult;

      return newResult;
    },
    key,
  });
}
```

```
export function selectorFamilyDeep<T, P extends SerializableParam>({
  get,
  key,
}: ReadOnlySelectorFamilyOptions<T, P>): (param: P) => RecoilValueReadOnly<T> {
  const cacheMap: Map<string, T | Promise<T> | RecoilValue<T>> = new Map();

  return selectorFamily({
    get: (args: P) => (getOptions) => {
      const newResult = get(args)(getOptions);
     // using internal stringify
      const mapKey = stringify(args, null, key);

      if (cacheMap.has(mapKey) && isEqual(cacheMap.get(mapKey), newResult)) {
        console.log(key, mapKey, ' is equal ', cacheMap.get(mapKey), newResult);

        return cacheMap.get(mapKey);
      }

      console.log(key, args, ' is NOT equal ', cacheMap.get(mapKey), newResult);

      cacheMap.set(mapKey, newResult);

      return newResult;
    },
    key,
  });
}
```

**Comment by david-trumid** [comment, marked as answer]
after reading up on the issues I see that the selector will still rerender down stream items however it does keep the object reference the same at least saving a child component rerender, cant wait for the value check option 👍

**Reply by BenjaBobs** [reply]
But the idea of being able to specify a custom comparison function is not a bad one.

```typescript
const deepEqSelector = selector({
	key: 'dEq',
    compare: (a,b) => {
		// deep equals comparison here
    	// falls back to reference equals if not specified
    },
	get: ({ get }) => {
		// data selection here
	}
})
```

**Reply by david-trumid** [reply]
my thoughts exactly

**Reply by drarmstr** [reply]
Yup, definitely a good idea.  The optimization to suppress re-rendering when a selector has the same reference-equality value is in the process of testing/landing.  The ability to suppress downstream selector evaluations and use custom comparators are  natural follow-ons.  Always wonderful if someone wants to help contribute the optimizations as well. ;)

**Reply by david-trumid** [reply]
sounds good, i will pull down repo this weekend and start digging in 👍

**Reply by dgrechka** [reply]
@drarmstr , is re-render suppression now enabled in recoil 0.3.1 for the cases when a selector has the same reference-equality value?

**Reply by drarmstr** [reply]
Not yet.  It is available with a feature flag if you want to enable `recoil_suppress_rerender_in_callback` in `Recoil_gkx.js` to test it.  There is currently a potential outstanding issue with Suspense that needs to be investigated.

**Reply by drarmstr** [reply]
Note this was released with 0.4.1
